### ISSUE TITLE
V2 - Validating int->str

### ISSUE THREAD

**Issue by vitalik** [pydantic/pydantic#6045]
Continue of the #5993

```Python
class Data(BaseModel):
    s: str
    f: str
    d: str


print(Data(s=42, f=4.2, d=Decimal('4.2')))
```

this gives validation error that `s f d - Input should be a valid string`

this was working in pydantic v1, and was actually very handy when dealing with parsing data of a public API where users send messy data from JS client sides

so would be nice to have some feature to be able to turn on str parsing (globally?)

### Affected Components

- [X] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)

**Update**: https://github.com/pydantic/pydantic/issues/6045#issuecomment-1704943954

**Comment by iBuitron** [comment]
It works fine for me.

But,

from pydantic import BaseModel

class User(BaseModel):
    id: int
    name = 'Jane Doe'
    
user = User(id='123')
user_x = User(id='123.45')

assert user.id == 123
assert user_x.id == 123
assert isinstance(user_x.id, int)

Gives me an Error:

`pydantic.error_wrappers.ValidationError: 1 validation error for User
id
  value is not a valid integer (type=type_error.integer)`
  
 It supposed to conver "123.45" to 123

**Comment by samuelcolvin** [comment]
But not many people have asked about it, the one person I've spoken to who mentioned this basically said "the new behaviour is right, the only thing that confused me was the documentation being wrong".

We'd need more demand for this before adding the shim.

**Comment by vitalik** [comment]
@samuelcolvin 

yeah, well that makes some sense, but my feeling is this will be pop up once it get released to wider audience

like imagine you used pydantic parse some yaml:

```Python
import yaml
from pydantic import BaseModel

YAML = """
details:
  name: 12345
  description: Hello world
 
"""


class Details(BaseModel):
    name: str
    description: str


raw_data = yaml.safe_load(YAML)
print(raw_data)

data = Details(**raw_data["details"])
```

I do not see any nice way to handle this except change ALL `str` field to some special string type

**Comment by dmontagu** [comment]
@vitalik just to be clear, we'd be able to get it to behave the old way (i.e., converting `int`s to `str`s, etc.) through just an annotation (i.e., changing the type hint from `str` to `Annotated[str, LenientStr()]` or something like that). Yes, you'd need to add the annotation everywhere in your code, but it would at least not be treated as a different type by type checkers. Not sure if that helps anything, just pointing it out since you said "some special string type" — not sure if your concern was that it would be a different type, or just that it would require changes everywhere (in the latter case, this doesn't help).

That said, I do still understand the interest in a "global" flag. (But I also understand Samuel's point about wanting to see more concrete demand first.)

**Comment by vitalik** [comment]
> but it would at least not be treated as a different type by type checkers

well that's my issue here - as you can see in my example i'm not initialising model with arguments (so type checkers do not play here) 

I'm passing raw yaml data just to parse and validate user input

Basically, I'm inspecting an impact of one of my clients where data parsing is heavily done with pydantic and migrating to v2 could simply lead to validation errors for data that was fine before

Changing all the code from `:str` to `:Annotated[str, force]` would be a really painful and annoying process, so it would be nice if you could provide some hook for easy monkey patching this behaviour (at least on a class level)

**Comment by dmontagu** [comment]
@adriangb I’m wondering if we could override the get_pydantic_core_schema or one of the other magic methods and have it insert annotations for certain types? I guess maybe a prepare pydantic annotations for a BaseModel class? Not sure if that would work on the class level instead of the field level but I’m thinking there might be something here that makes this “configurable” even without the use of config

**Comment by adriangb** [comment]
I think we can create or should be able to make changes so we can create a custom type like `LaxInt = Annotated[int, ...]`. The thing is that users would have to explicitly use that everywhere. I'm also not sure it would work right now, but we can (and should) re-factor internally to make it work.

We have also discussed about making a hook to do some sort of global replacement `{int: LaxInt}` but that has consequences for references and that sort of thing, so it probably needs a little bit more thought, even if it seems simple to implement.

**Comment by dmontagu** [comment]
Right I was mostly curious if there might be a way to hook into schema generation for models specifically as a way to handle this. I’ll play with it a bit this afternoon and see if I can find anything that works without code changes today, or at least see how far it is from working.

**Comment by adriangb** [comment]
So yeah you can do this:

```python
from typing import Annotated

from pydantic import BeforeValidator, TypeAdapter

LaxStr = Annotated[str, BeforeValidator(lambda x: str(x))]

ta = TypeAdapter(LaxStr)

print(ta.validate_python(123))
```

(Sorry about any confusion above about `LaxInt`)

But there's no way to set this globally or recursively for a model.

You could write a function that recursively inspects a `CoreSchema` and wraps any `{'type': 'str', ...}` schema with the validator above, then you could make that part of an annotation (`Annotated[T, RecursivelyMakeStrLax]`), a base class (`class LaxStrBaseModel(BaseModel): ... def __get_pydantic_core_schema__(...)`) or something like that.

**Comment by dmontagu** [comment]
Yeah that’s what I was thinking.

I think that might be enough even if it didn’t do any nested model recursion (though it might need to recurse on core schemas for the purpose of interacting properly with other validators etc., at which point maybe it’s easier to do it fully recursive) since the hypothetical user could just use that as the base class for all their models.

**Comment by vitalik** [comment]
> So yeah you can do this:

Caution
```Python
Annotated[str, BeforeValidator(lambda x: str(x))]

ta.validate_python(None) # will result -> "None" (as type str)
```

**Comment by dmontagu** [comment]
Probably want to use `pydantic.v1.validators.str_validator` or whatever the function was from v1

**Comment by mrenner-ML** [comment]
I am currently migrating v1 -> v2 and am facing issues with this. It would be great to at least have the option to set this in the model configs.

The use case I have for coercing int -> str is we often have IDs that appear as numericals (e.g. 12839384) and are infered as such when using pandas, whereas we want to treat them as string in order to perform string manipulation or join them with other non-numerical IDs. Pydantic v1 allowed me not to worry about this, whereas now I have to coerce all of my IDs separately beforehand.

May just add that I don't really see the benefit of removing this original behavior from v1, and that it will probably cause a lot of headaches when migrating, especially that this is not documented clearly.

**Comment by lig** [comment]
I think this could be another case for a type map we've discussed privately. As far as I remember one of the ideas was to have a global registry that will map an annotation to another one. It could look from a user point of view as something like this:

```py
type_map_registry: dict[type, type] = {str: Annotated[str, Field(force=True)]}
```

This would force all `str` fields in all modules to be treated as `Annotated[str, Field(force=True)]`.

**Comment by lig** [comment]
@mrenner-ML Indeed, this isn't stated explicitly in the migration guide. In general, coercing numbers and more complex types to `str` just hides errors and virtually disables the type checking. Overall, Pydantic V2 is more strict than Pydantic V1. However, there are more ways for customizing behavior in Pydantic V2.

In Pydantic V2 it's possible to achieve the desired behavior using [Custom Data Types](https://docs.pydantic.dev/dev-v2/usage/types/custom/).

**Comment by vitalik** [comment]
> It could look from a user point of view as something like this:
> 
> ```python
> type_map_registry: dict[type, type] = {str: Annotated[str, Field(force=True)]}
> ```

yeah, that should work for all of my cases... would also be nice to have this global and per base model class and a concrete model

**Comment by lig** [comment]
@vitalik I'm not sure when such a type map could be implemented. However, PRs are always welcome:)

**Comment by vitalik** [comment]
I can take a look, could you give some idea which module/class this best to add this functionality ?

**Comment by lig** [comment]
I think, `pydantic/main.py` is a good place to start. This is where `BaseModel` lives. However, it might lead to `pydantic/_internal/_model_construction.py` where `ModelMetaclass` is or to `pydantic/_internal/_generate_schema.py` where schema is being generated and actual schema variants are being chosen.

**Comment by vitalik** [comment]
with #6535 it would be possible to change annotation behaviour per model or even global 

```Python
BaseModel.model_config['replace_types'] = {str: LaxStr}
```

**Comment by vitalik** [comment]
@adriangb @dmontagu 

patching FieldInfo annotation does not really work...

see the following example - is this a bug or expected ?


```Python
from typing import Annotated
from pydantic.v1.validators import str_validator

LaxStr = Annotated[str, BeforeValidator(str_validator)]

class Data(BaseModel):
    gender: LaxStr = Field('F', max_length=1)


print(Data(gender='M'))
```
gives
```
gender
  Value should have at most 1 item after validation, not 1 [type=too_long, input_value='M', input_type=str]
    For further information visit https://errors.pydantic.dev/2.1.2/v/too_long
```

**Comment by LucasHartridge** [comment]
Hey guys :wave: 

I encounter the same issue migrating from v1 to v2. 

I am using Pydantic to build an XML later on so everything needs to be a String but also I need to support Ids to use SnowflakeIDs and Javascript can not handle them so I need to return them as strings

I know it can be painful to go on the code and update the str to the custom type but it is the only valid solution that I could find because I also agree that it should not be coerced anymore by default


```
from typing import Annotated

from pydantic import (
    PlainSerializer,
    WithJsonSchema,
)

# Represents an integer value that is serialized as a string
IntToStr = Annotated[
    int,
    PlainSerializer(lambda x: f"{x}", return_type=str),
    WithJsonSchema({"type": "string"}, mode="serialization"),
]

class MyAwesomePydanticSchema(BaseModel):
    id: IntToStr = Field(
        description="The corresponding ID",
        example=42,
        ge=1,
    )
```

Probably on the next iteration we can already include that on the library or add a property that can facilitate the usage

I am testing this at the moment an so far so good but if you have any opinion/place to improve let me know

**Comment by adriangb** [comment]
> @adriangb @dmontagu
> 
> patching FieldInfo annotation does not really work...
> 
> see the following example - is this a bug or expected ?
> 
> ```python
> from typing import Annotated
> from pydantic.v1.validators import str_validator
> 
> LaxStr = Annotated[str, BeforeValidator(str_validator)]
> 
> class Data(BaseModel):
>     gender: LaxStr = Field('F', max_length=1)
> 
> 
> print(Data(gender='M'))
> ```
> 
> gives
> 
> ```
> gender
>   Value should have at most 1 item after validation, not 1 [type=too_long, input_value='M', input_type=str]
>     For further information visit https://errors.pydantic.dev/2.1.2/v/too_long
> ```

I couldn't reproduce on `v2.0.3`:

```python
from typing import Annotated

from pydantic import BaseModel, BeforeValidator, Field
from pydantic.v1.validators import str_validator

LaxStr = Annotated[str, BeforeValidator(str_validator)]

class Data(BaseModel):
    gender: LaxStr = Field('F', max_length=1)
```

Runs without issues.

**Comment by adriangb** [comment]
@vitalik that example seems to work on `v2.0.3`:

```python
from typing import Annotated

from pydantic import BaseModel, BeforeValidator, Field
from pydantic.v1.validators import str_validator

LaxStr = Annotated[str, BeforeValidator(str_validator)]

class Data(BaseModel):
    gender: LaxStr = Field('F', max_length=1)


print(Data(gender='M'))
#> gender='M'
```

Given that we've offered a solution for this particular issue using the `LaxStr` approach, @vitalik can we close this and continue discussion of a `replace_types` like approach in #6535 ?

**Comment by vitalik** [comment]
well let's keep it open as it already added to milestone

**Comment by ehiggs** [comment]
Is there a follow up ticket to fix the documentation?

Currently [the page about strict mode](https://docs.pydantic.dev/latest/usage/strict_mode/) says:

>By default, Pydantic will attempt to coerce values to the desired type when possible. For example, you can pass the string "123" as the input to an int field, and it will be converted to 123. This coercion behavior is useful in many scenarios — think: UUIDs, URL parameters, HTTP headers, environment variables, user input, etc.

>However, there are also situations where this is not desirable, and you want Pydantic to error instead of coercing data.

>To better support this use case, Pydantic provides a "strict mode" that can be enabled on a per-model, per-field, or even per-validation-call basis. When strict mode is enabled, Pydantic will be much less lenient when coercing data, and will instead error if the data is not of the correct type.

How would it explain it better? A good smell for how a feature should work is to figure out the docs, and if the docs need to put in difficult to explain exceptions for how things work (like what's needed here) it likely means the exception is not a good idea. In this case, being unable to turn off strict mode for `str` seems like a weird thing to try to explain.

Perhaps:

>To better support this use case, Pydantic provides a "strict mode" that can be enabled on a per-model, per-field, or even per-validation-call basis. When strict mode is enabled, Pydantic will be much less lenient when coercing data, and will instead error if the data is not of the correct type. In our experiences, this is so important for parsing strings, pydantic will use strict on `str` types by default as an exceptional case. To disable this for strings, explicitly disable the strict setting on the field.

**Comment by lig** [comment]
@ehiggs Could you please submit a PR for the docs?

**Comment by alfaro28** [comment]
For anyone coming to this issue, as of version 2.1 this is the way to make it work for all models by default

```from pydantic import BaseModel, ConfigDict, GenerateSchema
from pydantic_core import CoreSchema, core_schema


class LaxStrGenerator(GenerateSchema):
    def str_schema(self) -> CoreSchema:
        return core_schema.no_info_plain_validator_function(str)


BaseModel.model_config = ConfigDict(schema_generator=LaxStrGenerator)
```

**Comment by adriangb** [comment]
Yes, but be careful monkey patching `BaseModel.model_config`. Obviously if more than one thing does it won't work.

**Comment by hampuskraft** [comment]
Hi, I'm not sure if this is intentional or not, but this doesn't seem to work with `pattern`. Here's an example:
```py
from pydantic import BaseModel, ConfigDict, Field, GenerateSchema
from pydantic_core import CoreSchema, core_schema


class LaxStrGenerator(GenerateSchema):
    def str_schema(self) -> CoreSchema:
        return core_schema.no_info_plain_validator_function(str)


class CustomBaseModel(BaseModel):
    model_config = ConfigDict(schema_generator=LaxStrGenerator)


class TestModel(CustomBaseModel):
    test: str = Field(min_length=1, max_length=32, pattern=r'^[a-zA-Z0-9 ]+$')
```
Which yields:
```
Traceback (most recent call last):
  File "/test.py", line 14, in <module>
    class TestModel(CustomBaseModel):
  File "/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 173, in __new__
    complete_model_class(
  File "/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 481, in complete_model_class
    cls.__pydantic_validator__ = SchemaValidator(simplified_core_schema, core_config)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.SchemaError: Invalid Schema:
model.schema.model-fields.fields.test.schema.function-after.pattern
  Extra inputs are not permitted [type=extra_forbidden, input_value='^[a-zA-Z0-9 ]+$', input_type=str]
    For further information visit https://errors.pydantic.dev/2.1/v/extra_forbidden
```
Thanks!

**Comment by adriangb** [comment]
Try `core_schema.no_info_before_validator(str, core_schema.str_schema())`

**Comment by hampuskraft** [comment]
I seem to have the same problem with that 🤔
```py
class LaxStrGenerator(GenerateSchema):
    def str_schema(self) -> CoreSchema:
        return core_schema.no_info_before_validator_function(str, core_schema.str_schema())
```

**Comment by adriangb** [comment]
This is indeed a bug/limitation. Working on it.

**Comment by wholmen** [comment]
Facing the same issue here when migrating from v1 to v2. Parsing huge excel sheets that I have little control over, and I want to convert some variables to string regardless of them being numbers or strings so I don't change any information when outputting in reports. 

Not clear what the optimal solution of the above options are. I would rather not go back to v1

**Comment by hampuskraft** [comment]
@wholmen Using `LaxStr = Annotated[str, BeforeValidator(lambda x: str(x))]` in place of `str`  in models where you want to cast int to str is the current solution, although you might also have luck with `LaxStrGenerator` above—unless you use `pattern`, until #6951 is merged.

**Comment by David-OConnor** [comment]
What is the way to do this in 2.3.0 with pattern? Thank you!

**Comment by 15498th** [comment]
There seems to be another instance of documentation describing old behavior: https://docs.pydantic.dev/latest/usage/types/standard_types/ (https://github.com/pydantic/pydantic/blob/main/docs/usage/types/standard_types.md?plain=1#L14):

> Strings are accepted as-is. `int`, `float`, and `Decimal` are coerced using `str(v)`.

**Comment by David-OConnor** [comment]
Solution: ` data_id: Union[int, str]`

**Comment by fortzi** [comment]
It would be great to have this as a configuration that pydantic loads from the environment/file.

The reason is that patching BaseModel this way has the disadvantage that it must be done before importing any module that uses BaseModel, which adds to the mental load and is easy to mess up without noticing.

Another possible solution is implementing a model validator in a base class that would replace BaseModel. This also has the disadvantage of having yet another thing to remember.

**Comment by 15498th** [comment]
@David-OConnor It might cause things break unexpectedly down the line if code using the model is not prepared to handle int. For example `print('Data id is:' + data.data_id)` raises TypeError if `data_id` value is integer.

**Comment by samuelcolvin** [comment]
I've re-opened this. On balance I think we should add a config flag to re-instate the V1 behavior since grander changes to allow type replacement is going to take much longer.

**Comment by CallumAtCarter** [comment]
I'm unsure if this is the same issue, but I'll post the issue I'm having here also. The data should be `str` throughout. If someone believes this is a different issue (or if I'm just doing this completely wrong) please let me know.

```py
StrippedStr = Annotated[
    str,
    BeforeValidator(lambda v: v.strip() if isinstance(v, str) else v),
]

AlphaStr = Annotated[
    StrippedStr,
    Field(pattern=r"^[a-zA-Z]+$"),
]


class Data(BaseModel):
    a: AlphaStr  # works
    b: AlphaStr | None  # does not work (see error below)
```

```
  File "site-packages\pydantic\_internal\_model_construction.py", line 173, in __new__
    complete_model_class(
  File "site-packages\pydantic\_internal\_model_construction.py", line 481, in complete_model_class
    cls.__pydantic_validator__ = SchemaValidator(simplified_core_schema, core_config)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.SchemaError: Invalid Schema:
model.schema.model-fields.fields.a.schema.default.schema.nullable.schema.function-before.pattern
  Extra inputs are not permitted [type=extra_forbidden, input_value='^[a-zA-Z]+$', input_type=str]
    For further information visit https://errors.pydantic.dev/2.1/v/extra_forbidden
```

**Comment by CallumWatkins** [comment]
@adriangb I should have updated before posting 😔. I will check on Monday and report back, thanks!

**Comment by CallumAtCarter** [comment]
@adriangb I should have updated before posting 😔. I will check on Monday and report back, thanks!

**Comment by berzi** [comment]
@adriangb I really need this feature but I'm not familiar with Pydantic's release schedule. When can I expect this to land in a release?

**Comment by lig** [comment]
As soon as the release is out 🤷🏻‍♂️ Usually, this is not long after a fix is ready.

At the moment, we are planning to make a release happen this week. There are a bunch of things coming to this release. Samuel outlined some priorities here https://github.com/pydantic/pydantic/issues/7324. This doesn't mean that everything from that list will make its way into the next release though.

**Comment by durgeksh** [comment]
Hi team,
I am new to Pydantic. My requirements are as below.
1. Check data type of coulmns
2. There are 5 columns out of which, I need to validate column 1 for first 500 rows only
3. Validate the data of one model's column with other model's column (Primary and Foreign key relation)
Thank you for the wonderful library.

**Comment by lig** [comment]
@durgeksh Hi! It seems like your comment is unrelated to the original issue. I'd recommend asking your question in a new discussion here https://github.com/pydantic/pydantic/discussions

**Comment by durgeksh** [comment]
@lig Thank you.
